### PR TITLE
JAMES-3754 IMAP SAVEDATE extension

### PR DESCRIPTION
**Pull request by quantranhong1999**

### CODE REVIEW

File: mpt/impl/imap-mailbox/jpa/src/test/java/org/apache/james/mpt/imapmailbox/jpa/JpaSelectedStateTest.java [hunk 1]
```diff
@@ -56,4 +56,9 @@ public void testUidKOREA() {
     @Override
     public void testUidUS() {
     }
+
+    @Override
+    public void testSearchSaveDate() {
+        // SEARCH save date just return naive/empty result for JPA.
+    }
```
Comment (chibenwa) [inline mpt/impl/imap-mailbox/jpa/src/test/java/org/apache/james/mpt/imapmailbox/jpa/JpaSelectedStateTest.java:63]: `@Ignore("SEARCH save date just return naive/empty result for JPA")`

Also, generally search indexes might need a delay between append and search to have the time to index data.

File: protocols/imap/src/main/java/org/apache/james/imap/encode/FetchResponseEncoder.java [hunk 2]
```diff
@@ -290,6 +296,19 @@ private void encodeInternalDate(ImapResponseComposer composer, FetchResponse fet
         }
     }
 
+    private void encodeSaveDate(ImapResponseComposer composer, FetchResponse fetchResponse) throws IOException {
+        final Optional<Date> saveDate = fetchResponse.getSaveDate();
+        if (isSaveDateFetched(saveDate)) {
```
Comment (vttranlina) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/FetchResponseEncoder.java:301]: We have Optional and still check null here?

Comment (quantranhong1999) [reply protocols/imap/src/main/java/org/apache/james/imap/encode/FetchResponseEncoder.java:301]: I needed a way to propagate whether the save date was fetched. With the current `FetchReponse`, the normal way is to use `null` to annotate it is not fetched.

However, the save date value is in Optional form, so it actually could make this a bit confusing, that is why I named `isSaveDateFetched(saveDate)` for more readability.

Better idea is welcome.

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Should be able to rebase i believe :)

**Pull request merged**